### ISSUE TITLE
Make it possible to access raw response body  

### ISSUE THREAD

**Issue by leonardehrenfried** [softwaremill/sttp#190]
In my program I want to log all outgoing requests - including their response bodies - but sttp's `Response` type only allows access to the already parsed type `T`. This greatly lessens the usefulness of logging everything for debugging purposes.

It would be great to have a way to read the raw response body even after it was being serialised.

I could imagine something like that:

```
val rawResponseBody: Array[Byte]
```

Not sure how streaming responses could be dealt with though.

Somewhat related to #66.

**Comment by guizmaii** [comment]
Hey @adamw,

I'm using the latest version (`2.2.3`) and can't find how to access this raw body. Was it removed?

**Comment by ghostbuster91** [comment]
@guizmaii 
Isn't `asByteArrayAlways`/`asByteArray` what you are looking for?
```
val response: Identity[Response[Array[Byte]]] = postEcho
        .body(testBody)
        .response(asByteArrayAlways)
        .send()
```

**Comment by adamw** [comment]
@guizmaii yes, there's no "raw body" accessible by default, as the raw body is usually a backend-specific single-use `Stream[Byte]` (for some stream implementation). I suppose one way to bypass this is to deserialize into a `WithRaw[T]` type (it's not in the sources, I just invented it ;) ) which would contain both low-level and high-level body representations.

What's your use-case for accessing the raw body, also logging?

**Comment by guizmaii** [comment]
> What's your use-case for accessing the raw body, also logging?

Yes because I have some serialisation error with `asJson[...]` and to debug them I have to log the raw response body (can't debug, sadly). It's not straightforward to transform a `.response(asJson[...])` to a logging `.response { raw -> ... }` and get the same result as with `asJson[...]` as the output of this function.

**Comment by adamw** [comment]
@guizmaii we might be able to do something better in sttp v3 (could you open an issue for logging the raw body? :) ), but meanwhile, maybe a specialised version of `asJson` would work. Taking circe (but all others work the same), we could do:

```scala
def asJsonLogged[B: Decoder: IsOption]: ResponseAs[Either[ResponseError[io.circe.Error], B], Any] =
  asString.mapWithMetadata { (body, meta) =>
    log.info(s"Received $body, status code: ${meta.code}")
    ResponseAs.deserializeRightWithError(deserializeJson)(body, meta)
  }
```

**Comment by ghostbuster91** [comment]
I am reopening this one because I think that we should make it possible to access the raw body for whatever reason and this issue is exactly about that.

Given that in v3 we already have `asEither` method it seems reasonable to have a symmetrical version - `asBoth`.
The usage might be following:
```
val response: Identity[Response[(UnderlyingRawType, Json)] = postEcho
        .body(testBody)
        .response(asBoth(asRaw, asJson))
        .send()
```
However, accessing the underlying raw type could be unsafe as it is likely to be bound to some resource. Because of that we might want to wrap it with some kind of Resource typeclass e.g. known from cats.

This fact brings also another limitation. Some of the native body sources like in okhttp can only be read once. As a consequence of that, the lowest possible representation of a body which we can return would be a base representation of a body in a particular sttp backend rather than raw type of underlying http client (e.g. in case of okHttp it will be `Publisher`[once we migrate to it] rather than `okhttp3.ResponseBody`)

**Comment by adamw** [comment]
Released in 3.0.0-RC1
